packages/flutter_tools/lib/src/widget_preview/analytics.dart

```diff
@@ -0,0 +1,89 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:unified_analytics/unified_analytics.dart';
+
+/// Manages analytics reporting for widget preview events sent from the Flutter Tool.
+///
+/// The generated widget preview scaffold project will also report analytics via DTD.
+class WidgetPreviewAnalytics {
+  WidgetPreviewAnalytics({required this.analytics});
+
+  final Analytics analytics;
+
+  final Stopwatch _launchTimer = Stopwatch();
+  final Stopwatch _reloadTimer = Stopwatch();
+
+  bool _launchIncludedProjectGeneration = false;
+
+  /// The analytics workflow for tracking widget previewer timings.
+  static const String kWorkflow = 'widget-preview';
+
+  /// The analytics event tracking widget previewer launch times.
+  static const String kLaunchTime = 'launch-time';
+
+  /// The analytics event tracking widget preview reload times.
+  static const String kPreviewReloadTime = 'preview-reload-time';
+
+  /// Provided as the label to [kLaunchTime] events if the widget preview scaffold project was
+  /// generated as part of the widget previewer starting up.
+  static const String kScaffoldGeneratedLabel = 'scaffold-generated';
+
+  /// Starts the stopwatch tracking the widget previewer launch time.
+  void initializeLaunchStopwatch() {
+    assert(!_launchTimer.isRunning);
+    _launchTimer.start();
+  }
+
+  /// Report that the current invocation of the `widget-preview start` command resulted in the
+  /// widget preview scaffold project being regenerated.
+  void generatedProject() => _launchIncludedProjectGeneration = true;
+
+  /// Send an analytics event reporting how long it took for the widget previewer to start.
+  void reportLaunchTiming() {
+    assert(_launchTimer.isRunning);
+    _launchTimer.stop();
+    analytics.send(
+      Event.timing(
+        workflow: kWorkflow,
+        variableName: kLaunchTime,
+        elapsedMilliseconds: _launchTimer.elapsedMilliseconds,
+        label: _launchIncludedProjectGeneration ? kScaffoldGeneratedLabel : null,
+      ),
+    );
+  }
+
+  /// Starts the stopwatch tracking the reload times for updated previews.
+  ///
+  /// This should be invoked when a file system event is detected in the preview detector.
+  void startPreviewReloadStopwatch() {
+    assert(!_launchTimer.isRunning);
+    _reloadTimer.start();
+  }
+
+  /// Send an analytics event reporting how long it took for a widget preview to be reloaded.
+  void reportPreviewReloadTiming() {
+    // TODO(bkonyi): only report when files are actually reloaded?
```

Comment:
@anderdobo, what do you think about this? Right now, we'll always report a preview reload timing event, even if the result is no files were hot reloaded into the preview environment (e.g., only whitespace was changed).

Even if a change results in no libraries being reloaded into the preview environment, we still need to do analysis and scan the updated file for changes to the previews defined within the file. That means there is still work being done that will also block processing of further changes to other files.

Reply:
LGTM given the context. Thanks!